### ISSUE TITLE
PSR12.Classes.ClassInstantiation incorrectly fix when using member vars and some variable formats

### ISSUE THREAD

**Issue by jrfnl** [squizlabs/PHP_CodeSniffer#2053]
WPCS already contained a `ClassInstantiation` sniff, so after seeing #2047 I've run the PSR12 version over the WPCS unit tests for the sniff to see if we could switch over at some point.

The test run identified a number of bugs in the PSR12 sniffs:

Test cases:
```php
$a = new self::$transport[$cap_string];
$renderer = new $this->inline_diff_renderer;
$a = new ${$varHoldingClassName};
```

Fixed as:
```php
$a = new self()::$transport[$cap_string];
$renderer = new $this()->inline_diff_renderer;
$a = new $(){$varHoldingClassName};
```

Expected fixes:
```php
$a = new self::$transport[$cap_string]();
$renderer = new $this->inline_diff_renderer();
$a = new ${$varHoldingClassName}();
```

:point_right:  Would you like me to look into fixing these ?

It also raises the question _Why does FIG ignore this rule for anonymous classes ?_, though that's outside the scope of PHPCS.

Other than the above, the WPCS sniff has some additional features.
- Detect class instantiation without parentheses in JS.
- Throw error when "_Assigning new by reference_" is detected (PHP only).
- Throw fixable error for whitespace between class name and open parenthesis (PHP + JS).
- Metrics for all errors.

:point_right:  Would you be interested in any of these ?

**Comment by gsherwood** [comment]
> Would you like me to look into fixing these ?

I'm working on this sniff right now due to another bug report, so I can take these ones. Thanks for reporting them.

> Would you be interested in any of these ?

They are good checks, but I don't think they belong in the PSR-12 sniff. I prefer these PSR ones to be as focused as possible due to the arguments that they tend to bring up.

**Comment by gsherwood** [comment]
I've fixed these cases. If you find others, please report new issues and I'll fix them up too.

**Comment by jrfnl** [comment]
@gsherwood Thanks for the quick turn-around. :heart: 

> They are good checks, but I don't think they belong in the PSR-12 sniff. I prefer these PSR ones to be as focused as possible due to the arguments that they tend to bring up.

So leave off ? or would you like me to pull these as separate sniffs in the `Generic` standard ?

**Comment by gsherwood** [comment]
> or would you like me to pull these as separate sniffs in the Generic standard ?

If you want to contrib them, and these checks don't exist elsewhere, then absolutely.
